pkg/util/winutil/pdhutil/pdhcounter.go

```diff
@@ -24,123 +29,222 @@ var (
 	if err != nil {
-		return nil, fmt.Errorf("Failed to make counter path %s: %v", counterName, err)
-	}
-	pdherror := pfnPdhAddEnglishCounter(p.query, path, uintptr(0), &p.counter)
-	if ERROR_SUCCESS != pdherror {
-		return nil, fmt.Errorf("Failed to add english counter %#x", pdherror)
+		counter.initError = fmt.Errorf("Failed to make counter path (\\%s(%s)\\%s): %v", counter.ObjectName, counter.InstanceName, counter.CounterName, err)
+		return counter.initError
 	}
-	pdherror = pfnPdhCollectQueryData(p.query)
+	pdherror := pfnPdhAddEnglishCounter(query.handle, path, uintptr(0), &counter.handle)
 	if ERROR_SUCCESS != pdherror {
-		return nil, fmt.Errorf("Failed to collect query data %#x", pdherror)
+		counter.initError = fmt.Errorf("Failed to add english counter (\\%s(%s)\\%s): %#x", counter.ObjectName, counter.InstanceName, counter.CounterName, pdherror)
+		return counter.initError
 	}
-	return &p, nil
+	counter.initError = nil
+	return nil
+}
+
+func (query *PdhQuery) AddCounter(counter PdhCounter) {
+	query.counters = append(query.counters, counter)
 }
 
-// GetEnglishSingleInstanceCounter returns a single instance counter object for the given counter class
-// the className and counterName must be in English.
+// AddEnglishCounterInstance returns a PdhSingleInstanceCounter that will fetch the value of a given instance of the given counter.
+// the objectName and counterName must be in English.
 // See PdhAddEnglishCounter docs for details
-// https://learn.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhaddenglishcountera
-func GetEnglishSingleInstanceCounter(className string, counterName string) (*PdhSingleInstanceCounterSet, error) {
-	return GetEnglishCounterInstance(className, counterName, "")
+// https://learn.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhaddenglishcounterw
+//
+// Implementation detail: This function does not actually call the Windows API PdhAddEnglishCounter. That happens
+//   when (*PdhQuery).CollectQueryData calls PdhCounter.AddToQuery. This function only links our pdhCounter struct
+//   to our PdhQuery struct.
+//   We do this so we can handle several PDH error cases and their recovery behind the scenes to reduce
+//   duplicate/error prone code in the checks that uses this package.
+//   For example, see tryRefreshPdhObjectCache().
+//   This function cannot fail, if it does then the checks that use it will need to be
+//   restructured so that their Configure() call does not fail if the counter is not available
+//   right away on host boot (see https://github.com/DataDog/datadog-agent/pull/13101).
+//   All errors related to the counter are returned from the GetValue()/GetAllValues() function.
+//
+func (query *PdhQuery) AddEnglishCounterInstance(objectName string, counterName string, instanceName string) PdhSingleInstanceCounter {
+	var p PdhEnglishSingleInstanceCounter
+	p.Initialize(objectName, counterName, instanceName)
+	query.AddCounter(&p)
+	return &p
```

Comment:
It feels like Go pater is more like 
p = NewPdhEnglishSingleInstanceCounter
query.AddCounter(&p)
return &p

I also wonder if pointer is that helpful here and may cause more performance harm than good in Go (from what I have read).  Of course for this impact it is like a hair splitting. I would check with go experts.